src/pu-package.c

```diff
@@ -0,0 +1,253 @@
+                child_size = g_file_info_get_size(child_info);
+                g_print(" (%s)", g_format_size(child_size));
+            }
+            g_print("\n");
+        }
+    }
+
+    return TRUE;
+}
+
+gboolean
+pu_package_list_contents(const gchar *package,
+                         gboolean print_size,
+                         GError **error)
+{
+    g_autofree gchar *dir_basename = NULL;
+    g_autofree gchar *mountpoint = NULL;
+    g_autofree gchar *layout_file = NULL;
+    g_autoptr(GFile) dir = NULL;
+
+    g_return_val_if_fail(g_strcmp0(package, "") > 0, FALSE);
+    g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+    if (!g_file_test(package, G_FILE_TEST_IS_REGULAR)) {
+        g_set_error(error, PU_PACKAGE_ERROR, PU_PACKAGE_ERROR_NOT_FOUND,
+                    "'%s' is not a regular file", package);
+        return FALSE;
+    }
+
+    dir_basename = g_strdup_printf("%s-%08x", PU_PACKAGE_BASENAME, g_random_int());
+    mountpoint = pu_create_mount_point(dir_basename, error);
+    if (!mountpoint)
+        return FALSE;
+
+    if (!pu_mount(package, mountpoint, "squashfs", "loop,ro", error))
+        return FALSE;
+
+    layout_file = pu_package_get_layout_file(mountpoint, error);
+    if (!layout_file) {
+        g_prefix_error(error, "Invalid package: ");
+        pu_umount(mountpoint, NULL);
+        return FALSE;
+    }
+
+    dir = g_file_new_for_path(mountpoint);
+    if (!print_dir_content(dir, TRUE, print_size, error)) {
+        pu_umount(mountpoint, NULL);
+        return FALSE;
+    }
+
+    if (!pu_umount(mountpoint, error))
+        return FALSE;
+
+    if (g_rmdir(mountpoint) < 0) {
+        g_set_error(error, PU_PACKAGE_ERROR, PU_PACKAGE_ERROR_FAILED,
+                    "Failed cleaning up directory '%s'", mountpoint);
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+gboolean
+pu_package_mount(const gchar *package,
+                 gchar **mountpoint,
+                 GError **error)
+{
+    g_autofree gchar *layout_file = NULL;
+
+    g_return_val_if_fail(*mountpoint == NULL, FALSE);
+    g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+    if (!g_file_test(package, G_FILE_TEST_IS_REGULAR)) {
+        g_set_error(error, PU_PACKAGE_ERROR, PU_PACKAGE_ERROR_NOT_FOUND,
+                    "'%s' is not a regular file", package);
+        return FALSE;
+    }
+
+    *mountpoint = g_strdup_printf("%s-%08x", PU_PACKAGE_PREFIX, g_random_int());
+    if (!pu_create_mount_point(g_path_get_basename(*mountpoint), error))
+        return FALSE;
+
+    if (!pu_mount(package, *mountpoint, "squashfs", "loop,ro", error))
+        return FALSE;
+
+    layout_file = pu_package_get_layout_file(*mountpoint, error);
+    if (!layout_file) {
+        g_prefix_error(error, "Invalid package: ");
+        pu_umount(*mountpoint, NULL);
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+gboolean
+pu_package_umount(const gchar *mountpoint,
+                  GError **error)
```

Comment:
Add input validation for `mountpoint` and `error`.

Reply:
Done.